ex/loqui/lib/loqui/client.ex

```diff
@@ -0,0 +1,358 @@
+  end
+
+  def handle_info({:close_go_away, go_away_code, go_away_data}, %State{sequence: :go_away, waiters: waiters}=state) do
+    # tell the waiters that the remote end went away, then close.
+    err = {:error, {:remote_went_away, go_away_code, go_away_data}}
+    Enum.each(waiters, fn {_, waiter} ->
+      Connection.reply(waiter, err)
+    end)
+
+    {:disconnect, err, state}
+  end
+
+  def handle_info(:send_ping, %State{sock: sock, last_ping: nil}=state) do
+    {next_seq, state} = next_sequence(state)
+
+    :gen_tcp.send(sock, Frames.ping(0, next_seq))
+    new_state = state
+      |> schedule_ping
+      |> Map.put(:last_ping, {next_seq, :erlang.system_time(:milli_seconds)})
+
+    {:noreply, new_state}
+  end
+  def handle_info(:send_ping, %State{last_ping: {sequence, _ts}}=state) do
+    {:stop, {:error, {:pong_not_received, sequence}}, state}
+  end
+
+  # Private
+
+  defp handle_packets([], state),
+    do: state
+  defp handle_packets([packet | rest], state) do
+    state = handle_packet(packet, state)
+    handle_packets(rest, state)
+  end
+
+  defp handle_packet({:ping, _flags, seq}, %State{sock: sock}=state) do
+    :gen_tcp.send(sock, Frames.pong(0, seq))
+    state
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{last_ping: {seq, _ts}}=state) do
+    %State{state | last_ping: nil}
+  end
+
+  defp handle_packet({:pong, _flags, seq}, %State{last_ping: nil, waiters: waiters}=state) do
+    new_waiters =
+      case Map.pop(waiters, seq) do
+        {nil, waiters} ->
+          Logger.error("Got an unknown pong for sequence #{inspect seq}")
+          waiters
+        {reply_to, waiters} ->
+          Connection.reply(reply_to, :pong)
+          waiters
+      end
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:hello_ack, _flags, ping_interval, data}, %State{}=state) do
+    [encoding, compression] = String.split(data, "|")
+    codec = Map.get(state.registered_codecs, encoding)
+    compressor = Map.get(state.registered_compressors, compression)
+
+    %State{state | codec: codec, compressor: compressor, ping_interval: ping_interval}
+      |> schedule_ping()
+  end
+
+  defp handle_packet({:response, _flags, sequence, payload}, %State{waiters: waiters, codec: codec}=state) do
+    {waiter, new_waiters} = Map.pop(waiters, sequence)
+    decoded_data = codec.decode(payload)
+    Connection.reply(waiter, decoded_data)
+
+    %State{state | waiters: new_waiters}
+  end
+
+  defp handle_packet({:go_away, _flags, close_code, payload_data}, %State{}=state) do
+    Process.send_after(self(), {:close_go_away, close_code, payload_data}, @go_away_timeout)
+    %State{state | sequence: :go_away}
+  end
+
+  defp handle_packet(packet, state) do
+    Logger.error("Received unknown packet, opcode #{inspect elem(packet, 0)} #{inspect packet}")
+    state
+  end
+
+  defp schedule_ping(%State{ping_interval: ping_interval}=state) do
+    Process.send_after(self(), :send_ping, ping_interval)
```

Comment:
We should send pings faster than the ping interval, right? Is `ping_interval` the max before it will time out?